src/core/graph.ts

```diff
@@ -392,45 +316,99 @@ export class DependencyGraph<T> {
     return "digraph G {" + dot + "\n}";
   }
 
-  public rootNodes(): Set<T> {
+  /**
+   * Return the nodes in the graph that have no upstream neighbors.
+   */
+  public getSourceNodes(): Set<T> {
     const roots = new Set<T>();
     /* Populate start set */
     for (const [v, e] of this.adjacencyMap) {
       if (e == null || e.size === 0) {
-        roots.add(v); // leaf nodes have no dependencies
-        // clone.delete(v); // FIXME add a removeNodes function to factor out the duplicate code below
+        roots.add(v);
       }
     }
     return roots;
   }
 
-  // Leaf nodes are nodes that do not depend on any other nodes.
-  //
-  // In the context of cyclic graphs it is therefore possible to
-  // have a graph without any leaf nodes.
-  // As a result, starting a graph search only from leaf nodes in a
-  // cyclic graph, will not necessarily traverse the entire graph.
-  public leafNodes(): Set<T> {
-    const leafs = new Set<T>(this.nodes());
-    for (const node of this.nodes()) {
-      for (const dep of this.getEdges(node)) {
+  /**
+   * Return the nodes in the graph that have no downstream neighbors.
+   */
+  public getSinkNodes(): Set<T> {
+    const leafs = new Set<T>(this.getNodes());
+    for (const node of this.getNodes()) {
+      for (const dep of this.getUpstreamNeighbors(node)) {
         leafs.delete(dep);
       }
     }
     return leafs;
   }
 }
 
-export class SortableDependencyGraph<
+/**
+ * A precedence graph with nodes that are sortable by assigning a numeric priority.
+ */
+export class SortablePrecedenceGraph<
   T extends Sortable<number>
-> extends DependencyGraph<T> {
-  updatePriorities(destructive: boolean, spacing = 100): boolean {
+> extends PrecedenceGraph<T> {
+  /**
+   * Create a sortable precedence graph. If a type and precedence graph are given,
+   * then remove all nodes that are not of the given type in a way the preserves
+   * the original lineage.
+   * @param type A type that extends T.
+   * @param pg A precedence graph.
+   */
+  constructor(
+    type?: new (...args: never[]) => T,
```

Comment:
@axmmisaka, I wasn't 100% sure about this change. It used to be `any[]` and this appears to work well without raising any eslint problems...

Reply:
I guess this is fine and intended, since we *never* assign to the arguments and this is meant to be a placeholder for "any argument"